File: ad_library/src/main/java/com/adgrowth/adserver/helpers/QueryStringHelpers.java
package com.adgrowth.adserver.helpers;

import android.net.Uri;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

import java.util.Iterator;
import java.util.Map;


public class QueryStringHelpers {
    public static String encode(JSONObject json) {
        StringBuilder result = new StringBuilder();
        encodeObject(result, "", json);
        return result.toString();
    }

    public static String encode(Map map) {
        StringBuilder result = new StringBuilder();
        encodeObject(result, "", new JSONObject(map));
        return "?" + result;
    }

    private static void encodeObject(StringBuilder result, String prefix, Object object) {
        try {
            if (object == null) return;
            if (object instanceof JSONArray) {
                JSONArray array = (JSONArray) object;
                for (int i = 0; i < array.length(); i++) {
                    try {
                        String newPrefix = prefix.isEmpty() ? i + "" : prefix + Uri.encode("[" + i + "]");
                        Object value = array.get(i);
                        encodeObject(result, newPrefix, value);
                        if (i + 1 < array.length() || value instanceof JSONObject || value instanceof JSONArray) {
                            result.append("&");
                        }
                    } catch (JSONException e) {
                        e.printStackTrace();
                        // throw new RuntimeException(e);
                    }
                }
            } else if (object instanceof JSONObject) {

                JSONObject json = (JSONObject) object;
                Iterator<String> keys = json.keys();
                while (keys.hasNext()) {
                    String key = keys.next();
                    String newPrefix = prefix.isEmpty() ? key : prefix + Uri.encode("[" + key + "]");
                    try {
                        Object value = json.get(key);
                        encodeObject(result, newPrefix, value);
                        if (keys.hasNext() || value instanceof JSONObject || value instanceof JSONArray) {
                            result.append("&");
                        }
                    } catch (JSONException e) {
//                        throw new RuntimeException(e);
                    }
                }
            } else {
                Object encodedValue = object;
                if (object instanceof String)
                    encodedValue = URLEncoder.encode(object.toString(), "UTF-8");
                result.append(prefix).append("=").append(encodedValue);

            }
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
            // throw new RuntimeException(e);
        }
    }
}

File: ad_library/src/main/kotlin/com/adgrowth/internal/integrations/adserver/helpers/QueryStringHelpers.kt
package com.adgrowth.internal.integrations.adserver.helpers

import android.net.Uri
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.UnsupportedEncodingException
import java.net.URLEncoder

/** This class does something like query-string-parser lib in nodejs does */
object QueryStringHelpers {
    fun encode(map: Map<*, *>?): String {
        val result = StringBuilder()
        encodeObject(result, "", map?.let { JSONObject(it) })
        return "?$result"
    }

    private fun encodeObject(result: StringBuilder, prefix: String, `object`: Any?) {
        try {
            if (`object` == null) return
            when (`object`) {
                is JSONArray -> {
                    for (i in 0 until `object`.length()) {
                        try {
                            val newPrefix =
                                if (prefix.isEmpty()) i.toString() + "" else prefix + Uri.encode(
                                    "[$i]"
                                )
                            val value = `object`[i]
                            encodeObject(result, newPrefix, value)
                            if (i + 1 < `object`.length() || value is JSONObject || value is JSONArray) {
                                result.append("&")
                            }
                        } catch (e: JSONException) {
                            // ignore
                        }
                    }
                }
                is JSONObject -> {
                    val keys = `object`.keys()
                    while (keys.hasNext()) {
                        val key = keys.next()
                        val newPrefix = if (prefix.isEmpty()) key else prefix + Uri.encode(
                            "[$key]"
                        )
                        try {
                            val value = `object`[key]
                            encodeObject(result, newPrefix, value)
                            if (keys.hasNext() || value is JSONObject || value is JSONArray) {
                                result.append("&")
                            }
                        } catch (e: JSONException) {
                            // ignore
                        }
                    }
                }
                else -> {
                    var encodedValue = `object`
                    if (`object` is String) encodedValue =
                        URLEncoder.encode(`object`.toString(), "UTF-8")
                    result.append(prefix).append("=").append(encodedValue)
                }
            }
        } catch (e: UnsupportedEncodingException) {
            // ignore
        }
    }
}
